### ISSUE TITLE
It should be an error to declare a structural type with multiple constructors of same structure

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#743]
Currently, this compiles:

```Haskell
type Day = Sun | Mon | Tue | Wed | Thu | Fri | Sat
```

It should give an error like: "This structural type has multiple constructors with the same structure. Try using `unique type`, and if this is your first time seeing this message you can learn more here: <url>."

Implementation should be straightforward. I'd probably have this happen before or as part of hashing of the data declarations in the file, which is done in FileParser.hs.

**Comment by runarorama** [comment]
I don't think this is sufficiently motivated. It's not obvious to why this should be an error.

Could this be a warning instead?

Do we really want to make it impossible to write types like `One | Two`?

**Comment by runarorama** [comment]
Note that there's an obvious workaround:

```
type Const a b = a

type Bool a b = Const () a | Const () b
```

**Comment by atacratic** [comment]
But if we do continue allowing structurally identical constructors, how do different people's incarnations of the type get reconciled?

I.e if I want my code using `type Bool = T | F` and your code using `type Boolean = False | True` (which are the same type according to hash) to interoperate, then how do I make sure to avoid getting True and False accidentally interchanged on the boundary?

**Comment by atacratic** [comment]
(suppose in the above that your code and my code is mostly already written, so it's not just a question of aliasing)

**Comment by runarorama** [comment]
@atacratic We could give a warning to the programmer. So if you defined `T | F`, we can warn you that the names `T` and `F` have no significance to Unison. We can also warn you (which we already do) if a type with identical structure already exists with different names.

**Comment by runarorama** [comment]
Yeah, so if I want to call your code where your convention and my convention have opposite meanings for `False` and `True` (maybe I called them `S` and `Ó`), then one of us would have to put the other's code into their codebase. And we can warn that different names exist for structurally identical constructors.

**Comment by runarorama** [comment]
I want to note very strongly that this is not an issue with structurally identical constructors. Check out the types I defined above, which have this exact issue but do not have structurally identical constructors.

**Comment by runarorama** [comment]
The really dangerous case is where two programmers have the exact same names but have chosen different semantics for those names. We can solve this by having a canonical order for constructor names. It might be weird to see `Fri | Mon | Sat | Sun | Thu | Tue | Wed`, but at least it would be consistent.

**Comment by pchiusano** [comment]
@runarorama those are some good examples! Makes me think it's kind of an impossible task to have a way of doing structural hashing that never produces confusing results and maybe it's weird to give an error in one case but not in the other cases. Or maybe giving an error is fine for cases we feel sure about - is there something wrong with doing that? 

I think it's at least valuable to give warnings, even if that's done on an ad hoc basis for some situations like this one that we happen to think of. And the easiest time to generate those warnings is when the user is writing the code - it's a purely local operation to check for this condition.

I don't love the idea of having any sort of verification pass on a whole codebase to check for confusing type aliasing situations, and this might need doing every time you pull in new code? I guess I have a lot of questions about how this idea would work exactly. If you can come up with a concrete proposal for that we could consider it though. :)

**Comment by runarorama** [comment]
One suggestion that came up in chat was to normalize all type expressions in data constructor declarations and error on constructors with (structurally) identical types. So `Fri` and `Mon` are identical because they're both `() -> Day`. So are `True (Trivial a)` and `False (Trivial b)` because `Trivial a = ()` and so they're both `() -> Bool`.

**Comment by pchiusano** [comment]
@runarorama I don't get your last comment. When would that sorting happen and what problem does that solve?

Also, do you object to having the original case (>1 constructor with identical type in a structural type) be an error? How about a warning? I take your point that even if it is an error or warning that does not fully guard against all possible confusing situations. That fact by itself doesn't seem like a compelling argument for or against having at least this one case be an error or warning - "we will only do this kind of verification if it prevents 100% of possible bugs".

**Comment by runarorama** [comment]
The sorting addresses the problem where identical constructor names have different assigned meanings.

We're already doing the whole-codebase guarding against different names for identical things. We show a message when a type already exists with a different name. We can easily detect this sort of thing on merge.

**Comment by runarorama** [comment]
I do object to the original case being an error, for the simple reason that it's really strange that you cannot define a 7-element type unless you jump through hoops.

It's also a little arbitrary that type constructor argument order matters, as does function argument order, but data constructor order for some reason does not.

**Comment by atacratic** [comment]
>  if I want to call your code where your convention and my convention have opposite meanings for False and True (maybe I called them S and Ó), then one of us would have to put the other's code into their codebase

That's undesirable though, right?  If you and I independently write some code, and then you can only call my code if you essentially fork and edit mine?  So I'm saying constructor misidentification is a genuinely bad thing we want to avoid.

>  So Fri and Mon are identical because they're both () -> Day. So are True (Trivial a) and False (Trivial b) because Trivial a = () and so they're both () -> Bool

This seems like a good improvement of the 'structurally identical constructors' check, because it handles these cases with phantom type parameters.

> `type Bool = Bool (Either () ())`

Not sure this is a great example because it just moves the question into the definition of `Either` - if the change covered by this issue went through then `Either` has to be `unique` I think.  So the `Bool` here becomes effectively unique because it hashes on `Either`'s GUID. 

> The really dangerous case is where two programmers have the exact same names but have chosen different semantics for those names. We can solve this by having a canonical order for constructor names

I think this probably is inherent to communication of all kinds!  It's insoluble except by describing things in terms of their extrinsic relations to other things.  

I think if you and I have defined data constructors named `Tue` but mean different things by them, then surely the type cannot be the canonical inhabitant of its structure class, so we should have made it unique.  

Also I don't see how imposing an order helps with the case where identical constructor names have different assigned meanings.  

> I do object to the original case being an error, for the simple reason that it's really strange that you cannot define a 7-element type unless you jump through hoops.

But seeing an error and then adding a `unique` like it tells you is not such a big hoop?  It also gives you valuable education on Unison's philosophy of type identity - that philosophy is more complicated than in other languages so it seems natural to me that it should sometimes manifest even in 'basic' type declarations.  

> It's also a little arbitrary that type constructor argument order matters, as does function argument order, but data constructor order for some reason does not.

It's weird but not arbitrary - type and function application is positional, but there's no analogous process for data constructors: for instance, you don't need your pattern matches to use the same order as the declaration.  I think this goes hand in hand with data constructors having names, whereas (alas 😉) function and type parameters don't.  

Other thoughts:

1.
I agree with partial warnings/errors being better than none at all.  But I think the improved idea for the 'structurally identical constructors' check might make this moot.  

2.
Did the phrase 'canonical inhabitant of its structure class' make sense?  My understanding/intuition is that you should use `unique` exactly when you can't say that your type is the canonical inhabitant of its structure class.  So to me it doesn't seem offensive that my `type Bool = True | False` has to be `unique` - because it has this internal permutation symmetry, i.e. there's no canonical boolean type, except up to quotient by this symmetry.  And we can't do the quotient for these purposes because we have to have a given constructor in hand and know which one we're talking about.  

I think this line of thinking takes you towards saying things like 'I can give an isomorphism between your Bool and my Boolean, so we can treat those `unique` types as interchangeable, and use that information to facilitate transparently calling between your code and mine, and when I'm doing `find : A -> Boolean` so I can find your functions too.  I'm not saying we want to start implementing that stuff right now, but I think it would be a decent direction.

**Comment by atacratic** [comment]
> Not sure this is a great example because it just moves the question into the definition of Either - if the change covered by this issue went through then Either has to be unique I think.

oops - disregard that bit, that was wrong... :-)

**Comment by pchiusano** [comment]
So to wrap this up, I think consensus is that we can at least make this case a warning / nice message for the user. It doesn't prevent every possible surprise and that is okay.

Separate from this ticket, we talked about making the `structural` vs `unique` modifier _required_, to bring some more awareness of the distinction and hopefully fewer surprises. 

And then we need to improve the docs, though how to improve them is still TBD.
